### ISSUE TITLE
Invalid object handles refreshing

### ISSUE THREAD

**Issue by bukka** [latchset/pkcs11-provider#494]
**Describe the bug**

Currently the pkcs11-provider is able to recreate a new session when it gets closed (implemented in `p11prov_get_session` and `slog_login` after calling the `session_check`). With the exception that it doesn't clear the errors, this works fine and the session gets recreated. The problem happens when some crypto operation (e.g. SignInit) is triggered on the key object that was created during the previous session. This object will use the object handle that no longer exist because the session is no longer existent. So the operation fails and there is no way to recover from this state (except restarting the application) so further operations fail as well.

**To Reproduce**
1. Start pkcs11-proxy daemon
2. Start application (e.g. nginx) using pkcs11-provider using pkcs11 proxy library connection to pkcs11-daemon 
3. Trigger some crypto operation (e.g. https request to nginx)
4. Restart pkcs11-proxy daemon (that clears all current sessions)
5. Send another crypto operation (this fails and any further requests fail as well due to using object handles attached to deleted session)

**Expected behavior**

Key objects should refresh when `CKR_OBJECT_HANDLE_INVALID` is returned and clear all errors if operation is successful.

**Operating environment (please complete the following information):**
 - OS: Ubuntu 20.04 but with custom build of OpenSSL 3.4

**Token and application used (please complete the following information):**
 - [pkcs11-proxy](https://github.com/bukka/pkcs11-proxy) proxying to SoftHSMv2

**Additional context**

My application uses pkcs11-provider with nginx that checks ssl errors so it is also important that all errors are clearer. The whole setup is a bit complex but I can provide more info if anyone needs to recreate it. I have actually already debug everything and have an initial fix so I'm going to create a PR shortly. If we agree on the approach I would look into some automated tests so not putting all details how to manually recreate it here yet.

**Comment by bukka** [comment]
Draft of the fix proposed in #495

**Comment by simo5** [comment]
This is more complex to deal with than it appears, while you could try to renew handles one by one, this is something that should somehow detected and corrected for all objects at the same time, otherwise you may have tokens that on restart reuse handles, and then you get refreshes coming in out of order, such that the first few "invalid" handles are reported as such, and then later a "valid" but incorrect handle ends up used ie you may have the incorrect keys used because in run one it had handles 42 but in run two handle 42 is associated with a different key ...

**Comment by simo5** [comment]
Also if the proxy is restarted, wouldn't any session operation fail given the module is not re-initialized ?

I think it is the client pkcs#11 driver duty to either completely shut down or it has to re-establish handles ... somehow.

What we could do is try to keep a mapping of session to ephemeral handles and if that session ever returns back as invalid, we would invalidate all handles.

But for things like private keys this will just delay the inevitable failure ...

**Comment by bukka** [comment]
So I manage to make this work for that limited use in my fork ( https://github.com/bukka/pkcs11-provider/commits/nginx-pkcs11-proxy-v2/ ) that my client uses. This is based on older version because rebases had been quite painful so I kind of froze it there and would like to upstream as much as we can - ideally having working version without any patches.

The way how it works is that it needs to do reinit (which is basically what is done after the fork) after there is a device error ( https://github.com/bukka/pkcs11-provider/commit/0caf09a762ab600be70f51f36027fed97ac16298 and related fix https://github.com/bukka/pkcs11-provider/commit/1e234b94354516c6754d2bfb3cd7594411322a48 ) and then do the refresh ( https://github.com/bukka/pkcs11-provider/commit/f266b964dfe7fac90daf15b26ed01055946042fa  which is what is in #495 ).

I realise that always doing re-init after device error should not be done by default as some application might not want to do it but I was wondering if you would be fine to have this optional through some configuration options?

The second part (the refresh on error) makes sure that the outdated handles get refreshed because relaying on full re-init still had space for errors which happened as soon as there was some traffic (IIRC it was when the restarted happened between verify init and update but not sure as it has been some time).

**Comment by simo5** [comment]
Yes I am willing to entertain a quirk that allows "re-init after errors", but we'd need to define what kind of error will trigger this.

As far as "space for errors", in a recent PR we landed a better locking mechanism for the init section that I think we can use to do a full invalidation of all handles. I am ok letting refreshes happen as handles are used so no unnecessary refresh is done at re-init, but the handles should all be invalidated to avoid the risk of handle confusion, which is a security issue.

**Comment by bukka** [comment]
In my case the error is always CKR_DEVICE_ERROR so that would be the trigger. 

 I looked into 8571b6f which I assume is the change that you are talking about. It should help me to make the re-init serialized and invalidate all handles in one place so that should cover the handle confusion. It won't help with the operation that actually hits the error though so I think I will still need the fallback for that one but it would do the full invalidation and retry instead of refreshing just the object as in #495. Multipart operation that is already past init will fail either way which I think is fine.

Once that other issue is resolved I will look into the patch for this one and see what can be done.
